src/client/cache.ts

```diff
@@ -0,0 +1,37 @@
+export class SuperCache<T> {
+  private cache: Record<string, T> = {};
+
+  getCached(cacheKey: string, initializer: () => T): T;
+  getCached(cacheKey: string, initializer: () => Promise<T>): Promise<T>;
+  getCached(
+    cacheKey: string,
+    initializer: () => T | Promise<T>
+  ): T | Promise<T> {
+    const cached = this.cache[cacheKey];
+    if (cached !== undefined) {
+      return cached;
+    }
+
+    const initialized = initializer();
+    if (initialized instanceof Promise) {
+      return new Promise((resolve, reject) => {
+        initialized
+          .then(value => {
```

Comment:
question: Does this have to create a new Promise? Can't it just be
```
initialized.then(value => {
  this.cache[cacheKey] = value;
  return value;
})
```
?

Reply:
good question